### ISSUE TITLE
Include scripts-styleguide.js by default in library definitions

### ISSUE THREAD

**Issue by rupl** [fourkitchens/emulsify#68]
I was stuck trying to figure out why I couldn't see a concatenated file that seemed to be generating properly when I realized there are two JS tasks in the gulp file. One is for global JS and that is already defined in `emulsify.libraries.yml` but the other is `scripts-styleguide.js` which concats all the individual components and makes them available to Drupal.

A simple addition to the YML got me up and running:

```yml
  js:
    dist/scripts.js: {}
    dist/scripts-styleguide.js: {}
  dependencies:
    - core/jquery
```

**Comment by evanmwillhite** [comment]
@rupl hey, thanks for using the project! You could definitely take this approach above if your JS is minimal, but we typically only include component js files in Drupal as needed. Here's [a link to the section in the wiki on it](https://github.com/fourkitchens/emulsify/wiki/Drupal-Components#javascript-in-drupal). 

tl;dr is that D8 gives us this handy Twig function: `{{ attach_library('emulsify/LIBRARY_NAME') }`. Just add a new library for each component in libraries.yml (example at bottom [here](https://github.com/fourkitchens/emulsify/blob/master/emulsify.libraries.yml)) and then use that attach_library function in whatever Drupal Twig file needs the JS. Pretty sweet!

**Comment by rupl** [comment]
Ok that makes sense. I have another question in that case. I reconfigured my component as you advised, but when I run Emulsify's build/watch task, I see the following error when it attempts to build PL:

```
PHP Fatal error:  Uncaught Twig_Error_Syntax: Unknown "attach_library" function. in /PATH/TO/web/themes/custom/emulsify/components/_patterns/03-organisms/site/nav-primary/nav-primary.twig:1
```

I see on [api.drupal.org](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Template%21TwigExtension.php/function/TwigExtension%3A%3AattachLibrary/8.2.x) that `attach_library` is Drupal-specific, not part of "standard" Twig which Pattern Lab is using. I don't see any reference to a Twig Extension in the Emulsify config files. I did notice there are kits other than `pattern-lab/edition-twig-standard` which is what Emulsify set me up with.

Did I install or initialize Emulsify incorrectly? Or am I missing where this config needs to be set?

**Comment by evanmwillhite** [comment]
`attach_library` is Drupal-specific - sorry I wasn't more clear on where that goes. That goes into the Drupal Twig file - whichever file you're using your `include`,`extends` or `embed` function in to pull in your component markup (in the `/templates` dir if you're using the default approach).  Pattern Lab gets all js loaded via that scripts-styleguide.js file (performance not being as important there), so that function is only needed on the Drupal side.

That said, it's good to know that if you need to put any Drupal-specific Twig function (or filter, tag, etc.) and have PL recognize it, you can. Those go into /_twig-components. There's a few examples in there you can use as a starting point.

**Comment by rupl** [comment]
ah I see what I did now, I put the call in the PL component and it belongs in the `templates` folder with the Drupal-specific `.html.twig` file

**Comment by adamfchs** [comment]
I have a question on getting jquery to work within pattern-lab and pathing to core Drupal assets from _meta/_01-foot.twig:
There are two Drupal specific files like this: <script src="../../../../../../../core/misc/drupal.js"></script>
I'm assuming if I can get those working than the jquery function that wraps my Drupal behaviors will not through Jquery is not defined errors.

The issue is that no matter what path I try to use for ../../../../../../../core/misc/drupal.js, I keep getting file not found for that file.  I've tried going from /core.... to ../../../../../../../../../../../../../../../.... one step at a time and the drupal js file has never been found.

I'm using a site with a composer /web root  and from there the drupal.ls path is /web/core/misc/drupal.js—so that part is correct.  My styleguide.html file is /web/themes/themename/pattern-lab/public/styleguide/html/styleguide.html

Any idea why that core drupal file is not being found from the patternlab footer?  And or any other idea why this script would work in the drupal site but throws jquery not defined errors in pattern lab?

(function ($) {
Drupal.behaviors.searchToggle = {
	attach: function (context, settings) {
	 	$(".block-search-form-block", context).once('tSearch').each(function(){  
      //insert toggle icon
      $(this).prepend('<span class="t-search">Search</span>');
	 		$('.t-search', this).click(function(){
        $(this).toggleClass('searching');
	 			$('#search-block-form').toggleClass('slideDown');
	 		});	
		});
	}
}
})(jQuery);

**Comment by rupl** [comment]
One method that I've had success with is symlinking the necessary files into the PL directory. For instance on a recent project I needed three files. I symlinked them to `components/js` and committed the symlink to version control.

```
cd web/themes/custom/emulsify/components/js
ln -s ../../../../../core/misc/drupal.js ./
```

The above code I pulled from my bash history, your commands might require different directory depths but the idea is the same. Now the the PL build step naturally copies these files into the generated JS output for Pattern Lab.  Since the JS is now accessible inside the PL public directory, you can reference them using a much shorter relative path inside `_01-foot.twig` template:

```diff
+  <script src="../../js/ready.min.js"></script>
-  <script src="../../../../../../../core/misc/drupal.js"></script> -->
+  <script src="../../js/drupal.js"></script>
+  <script src="../../js/debounce.js"></script>
```

The end result is that you can author all your JS as if you are inside Drupal and it works in both places. No more Drupal-specific errors inside Pattern Lab!

**Comment by adamfchs** [comment]
Thanks so much for the symlink approach Rupl!  Worked like a charm to remove the errors.

One other question I have is that I use the Drupal behaviors and context approach to limit the scripts to only where they are needed (along with attach_library).

I've got the search toggle script above working in both places, but only after I remove:
Drupal.behaviors.searchToggle = {
	attach: function (context, settings) {
	 	$(".block-search-form-block", context).once('tSearch').each(function(){

My PL and drupal search block both have the class .block-search-form-block so PL should be finding that context—I think that for some reason its not recognizing the Drupal.behaviors part.  

Do you use Drupal.behaviors? Have you run into this snag in PL? I've symlinked all js files listed as dependencies in the my libraries.yml.

**Comment by evanmwillhite** [comment]
Kudos @rupl - great idea with the symlinks. As for making behaviors work outside Drupal, as long as all dependencies are correctly loaded now, @adamfchs I've had luck adding this below my other scripts in `_01-foot.twig`: `<script>Drupal.attachBehaviors();</script>`. The reason I add this there is because Drupal takes care of calling that at appropriate times (on load, after AJAX calls) so it doesn't need that code in Drupal. Hopefully that helps!

**Comment by adamfchs** [comment]
Thanks @evanmwillhite !!!  Attaching the behaviors in the footer.twig in pl worked perfectly.  All scripts are now running in both places seamlessly.

**Comment by evanmwillhite** [comment]
Yay!!

**Comment by shanejeffers** [comment]
When I run the build process with the symlinks, pattern lab doesn't copy them into the public folder.  Any ideas why that would happen?
